### PR TITLE
initial implementation of event hooks

### PR DESCRIPTION
**Pull request by dylanlott**
Callback implementation of a pub-sub system.

### CODE REVIEW

**Comment by dylanlott** [comment]
_*screams in callback*_

File: pkg/kademlia/kademlia.go [hunk 1]
```diff
@@ -168,13 +173,19 @@ func (k *Kademlia) GetRoutingTable(ctx context.Context) (dht.RoutingTable, error
...
 	if len(k.bootstrapNodes) == 0 {
 		return BootstrapErr.New("no bootstrap nodes provided")
 	}
 
-	return k.lookup(ctx, k.routingTable.self.Id, discoveryOptions{
+	err := k.lookup(ctx, k.routingTable.self.Id, discoveryOptions{
 		concurrency: k.alpha, retries: defaultRetries, bootstrap: true, bootstrapNodes: k.bootstrapNodes,
 	})
+
+	if err != nil {
```
Comment (jtolds) [inline pkg/kademlia/kademlia.go]: nit: why this change?

File: pkg/signal/main.go [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+	"storj.io/storj/pkg/utils"
+)
+
+// ErrNilCallback throws if there is no callback function provided
+var ErrNilCallback = errs.Class("must provide a callback function to register")
+
+// Dispatcher is what's loaded onto each service
+type Dispatcher struct {
+	sync.RWMutex
```
Comment (jtolds) [inline pkg/signal/main.go]: It doesn't look like you need an RWMutex over a Mutex. If all you need is a Mutex, it would be more efficient to just use it.

File: pkg/signal/main.go [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+	source    string
```
Comment (jtolds) [inline pkg/signal/main.go]: It doesn't look like you use source for anything.

File: pkg/signal/main.go [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+	}
+}
+
+// Dispatch broadcasts an event across all callbacks
+// If any of them error, it will return that error,
+func (d *Dispatcher) Dispatch(name string) error {
+	d.Lock()
+	defer d.Unlock()
+
+	errors := []error{}
```
Comment (jtolds) [inline pkg/signal/main.go]: as soon as utils.ErrorGroup is in, you can use it for this: https://github.com/storj/storj/pull/636/files#diff-fb81b0a96eed5b010cfb081b3aef4431R82

File: pkg/signal/main.go [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+	if d.callbacks[name] == nil {
+		d.callbacks[name] = make([]Callback, 0)
+	}
+
+	for _, cb := range d.callbacks[name] {
```
Comment (jtolds) [inline pkg/signal/main.go]: all of these callbacks are happening under a lock. if any of these callbacks call Register, we will deadlock. can we make a copy of all of the callbacks first, unlock, then call all of them?

File: pkg/signal/main.go [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+// Dispatch broadcasts an event across all callbacks
+// If any of them error, it will return that error,
+func (d *Dispatcher) Dispatch(name string) error {
+	d.Lock()
+	defer d.Unlock()
+
+	errors := []error{}
+
+	if d.callbacks[name] == nil {
+		d.callbacks[name] = make([]Callback, 0)
```
Comment (jtolds) [inline pkg/signal/main.go]: safe to say you could return early here

File: pkg/signal/signal_test.go [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+	assert.Equal(t, err, errors.New("error"))
+
+	err = d.Register("test:normal", func() error {
+		return nil
+	})
+	err = d.Dispatch("test:normal")
+	assert.Equal(t, err, nil)
+
+	err = d.Register("test:multiple", func() error {
+		return nil
```
Comment (jtolds) [inline pkg/signal/signal_test.go]: do you want any of your tests to do something in the callback, e.g., increment a counter or something, so you know it did stuffs?

File: pkg/signal/main.go [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+func NewDispatcher(source string) *Dispatcher {
+	return &Dispatcher{
+		source:    source,
+		callbacks: make(map[string][]Callback),
+	}
+}
+
+// Dispatch broadcasts an event across all callbacks
+// If any of them error, it will return that error,
+func (d *Dispatcher) Dispatch(name string) error {
```
Comment (jtolds) [inline pkg/signal/main.go]: the string type might actually be bad for the "name" value. the reason is due to namespace collisions. let's say you have two packages, and they both use the same dispatcher, and accidentally both use the `name` "updates". you're going to have some really subtle bugs that you could actually enforce the compiler to check for you!

so there are (at least) two options. option 1: 

this is going to sound straight up bat crazy, but an interface{} may actually be a better choice for a name type here. each package can define it's own type and use a value of that type. because each type is from a different package, they will never be equal, and never show up in the map as the same key. voila, the compiler prevents this entire class of bug.

e.g.:

```
type myDispatchType int

func Whatever() {
  ...
  err = d.Dispatch(myDispatchType(1))
  ...
```

option 2:

another way of dealing with this is to make a GenSym() function (generate symbol) and only accept symbols. like this:

```
type Symbol uint64

var symbolCounter = new(uint64)

func GenSym() Symbol {
  return Symbol(atomic.AddUint64(symbolCounter, 1))
}

func (d *Dispatcher) Dispatch(name Symbol) error {
   ...
```

then each package would do this:

```
var (
  UpdateChannel = signal.GenSym()
)

func Whatever() {
  ...
  err = d.Dispatch(UpdateChannel)
  ...
```
